File: src/MagnusAddon/note/sentences/word_exclusion_set.py
from __future__ import annotations

from typing import TYPE_CHECKING, override

from autoslot import Slots
from language_services.janome_ex.word_extraction.word_exclusion import WordExclusion
from typed_linq_collections.collections.q_unique_list import QUniqueList

if TYPE_CHECKING:
    from collections.abc import Callable

class WordExclusionSet(Slots):
    def __init__(self, save_callback: Callable[[], None], exclusions: list[WordExclusion]) -> None:
        self._save: Callable[[], None] = save_callback
        self._exclusions: QUniqueList[WordExclusion] = QUniqueList(exclusions)
        self._excluded_words: QUniqueList[str] = self._extract_words()

    def get(self) -> QUniqueList[WordExclusion]: return self._exclusions

    def _extract_words(self) -> QUniqueList[str]:
        return QUniqueList(exclusion.word for exclusion in self._exclusions)

    def words(self) -> QUniqueList[str]:
        return self._excluded_words

    def reset(self) -> None:
        self._exclusions.clear()
        self._excluded_words.clear()
        self._save()

    def add_global(self, vocab: str) -> None:
        self.add(WordExclusion.global_(vocab))

    def add(self, exclusion: WordExclusion) -> None:
        self._exclusions.add(exclusion)
        self._excluded_words = self._extract_words()
        self._save()

    def remove(self, exclusion: WordExclusion) -> None:
        self._exclusions.remove(exclusion)
        self._excluded_words = self._extract_words()
        self._save()

    def remove_string(self, to_remove: str) -> None:
        for exclusion in [ex for ex in self._exclusions if ex.word == to_remove]:
            self._exclusions.remove(exclusion)
        self._excluded_words = self._extract_words()
        self._save()

    def excludes_at_index(self, word: str, index: int) -> bool:
        return any(exclusion for exclusion in self._exclusions if exclusion.excludes_form_at_index(word, index))

    @override
    def __repr__(self) -> str: return ", ".join(exclusion.__repr__() for exclusion in self._exclusions)

File: src/src_dotnet/JAStudio.Core/Note/Sentences/WordExclusionSet.cs
using System;
using System.Collections.Generic;
using System.Linq;
using JAStudio.Core.LanguageServices.JanomeEx.WordExtraction;

namespace JAStudio.Core.Note.Sentences;

public class WordExclusionSet
{
    private readonly Action _saveCallback;
    private readonly HashSet<WordExclusion> _exclusions;
    private HashSet<string> _excludedWords;

    public WordExclusionSet(Action saveCallback, List<WordExclusion> exclusions)
    {
        _saveCallback = saveCallback;
        _exclusions = new HashSet<WordExclusion>(exclusions ?? new List<WordExclusion>());
        _excludedWords = ExtractWords();
    }

    public static WordExclusionSet Empty()
    {
        return new WordExclusionSet(() => { }, new List<WordExclusion>());
    }

    public static WordExclusionSet Empty(Action saveCallback) => new(saveCallback, new List<WordExclusion>());

    public bool IsEmpty() => !_exclusions.Any();

    public IEnumerable<WordExclusion> Get() => _exclusions;

    private HashSet<string> ExtractWords()
    {
        return _exclusions.Select(e => e.Word).ToHashSet();
    }

    public HashSet<string> Words() => _excludedWords;

    public void Reset()
    {
        _exclusions.Clear();
        _excludedWords.Clear();
        _saveCallback();
    }

    public void AddGlobal(string vocab)
    {
        Add(WordExclusion.Global(vocab));
    }

    public void Add(WordExclusion exclusion)
    {
        _exclusions.Add(exclusion);
        _excludedWords = ExtractWords();
        _saveCallback();
    }

    public void Remove(WordExclusion exclusion)
    {
        _exclusions.Remove(exclusion);
        _excludedWords = ExtractWords();
        _saveCallback();
    }

    public void RemoveString(string toRemove)
    {
        var toRemoveList = _exclusions.Where(ex => ex.Word == toRemove).ToList();
        foreach (var exclusion in toRemoveList)
        {
            _exclusions.Remove(exclusion);
        }
        _excludedWords = ExtractWords();
        _saveCallback();
    }

    public bool ExcludesAtIndex(string word, int index)
    {
        return _exclusions.Any(exclusion => exclusion.ExcludesFormAtIndex(word, index));
    }

    public override string ToString()
    {
        return string.Join(", ", _exclusions.Select(e => e.ToString()));
    }
}
